File: android/app/src/main/kotlin/pro/kaprovpn/android/core/Subscription.kt
package pro.kaprovpn.android.core

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.HttpURLConnection
import java.net.InetSocketAddress
import java.net.Proxy
import java.net.Socket
import java.net.URI
import java.util.Base64
import javax.net.ssl.SSLException

/**
 * Импорт подписок — порт `kapro_vpn/core/subscription.py`.
 *
 * Большинство VPN-провайдеров (dns.army, BMV, AmneziaFree, ...) выдают
 * один URL, который возвращает base64-encoded список share-URL'ов — по
 * одному на строку. Этот модуль скачивает тело, декодирует, парсит
 * каждый share-URL через [ShareUrlParser] и возвращает готовый
 * [SubscriptionResult].
 *
 * Format detection (как в десктопе):
 * - Если тело содержит явные share-URL (vless://, vmess://, ...) →
 *   парсим как plain.
 * - Иначе пробуем base64-decode; если декод даёт scheme'ы — используем.
 *
 * DPI-fallback (десктоп): можно фетчить через локальный xray-tunnel
 * если direct-request заблокирован DPI. На Android пока не нужно —
 * пользователь может сначала включить VPN и потом импортнуть.
 */
object Subscription {

    private val SUPPORTED_SCHEMES = listOf(
        "vless://", "vmess://", "trojan://", "ss://",
        "hysteria2://", "hy2://",
    )

    private const val USER_AGENT =
        "KaproVPN-Android/0.1.0-dev (Android; +https://github.com/fafnirov/KaproVPN)"

    private const val CONNECT_TIMEOUT_MS = 10_000
    private const val READ_TIMEOUT_MS = 20_000

    /**
     * Результат загрузки и парсинга подписки.
     *
     * @param configs успешно распарсенные ProxyConfig
     * @param errors короткие human-readable строки об ошибках (по одной
     *   на каждый share-URL, который не распарсился)
     * @param rawLines сколько кандидатов мы попытались распарсить —
     *   полезно для UI «5 из 7 распарсилось»
     */
    data class Result(
        val configs: List<ProxyConfig>,
        val errors: List<String>,
        val rawLines: Int,
        /** true если fetch прошёл через локальный xray-туннель (DPI-fallback).
         *  Используется UI чтобы показать «загружено через VPN». */
        val viaProxy: Boolean = false,
    )

    /**
     * Вытаскивает share-URL'ы из тела ответа подписки.
     *
     * Сначала проверяет нет ли в теле уже plain share-URL'ов. Если нет —
     * пробует base64-decode (с автоматическим fix padding'а). Возвращает
     * пустой список если ничего извлечь не удалось.
     */
    fun parseBody(body: String): List<String> {
        val trimmed = body.trim()
        if (trimmed.isEmpty()) return emptyList()

        // Сначала пробуем plain. Если в теле нет ни одного известного
        // scheme — попробуем base64.
        val candidates = mutableListOf(trimmed)
        if (SUPPORTED_SCHEMES.none { sch -> sch in trimmed }) {
            try {
                // Fix padding до длины кратной 4 (как в parser._b64DecodePadded)
                val padded = trimmed + "=".repeat((-trimmed.length).mod(4))
                val normalized = padded.replace('-', '+').replace('_', '/')
                val decoded = Base64.getDecoder().decode(normalized)
                    .toString(Charsets.UTF_8)
                if (SUPPORTED_SCHEMES.any { sch -> sch in decoded }) {
                    // base64 победил → используем его как primary candidate
                    candidates.add(0, decoded)
                }
            } catch (_: Exception) {
                // Не base64 — оставляем только plain
            }
        }

        for (text in candidates) {
            val urls = text.lines()
                .map { it.trim() }
                .filter { line ->
                    line.isNotEmpty() &&
                        !line.startsWith("#") &&
                        SUPPORTED_SCHEMES.any { sch -> line.startsWith(sch) }
                }
            if (urls.isNotEmpty()) return urls
        }
        return emptyList()
    }

    /**
     * Парсит body и собирает результат. Делится с `import` чтобы можно
     * было использовать на manually-pasted body (когда сервер блокирует
     * запросы из приложений — пользователь копирует тело из браузера).
     */
    fun resultFromBody(body: String, viaProxy: Boolean = false): Result {
        val shareUrls = parseBody(body)
        val configs = mutableListOf<ProxyConfig>()
        val errors = mutableListOf<String>()
        for (url in shareUrls) {
            try {
                configs.add(ShareUrlParser.parse(url))
            } catch (e: ParseError) {
                val short = url.take(60) + (if (url.length > 60) "…" else "")
                errors.add("$short — ${e.message}")
            }
        }
        return Result(
            configs = configs,
            errors = errors,
            rawLines = shareUrls.size,
            viaProxy = viaProxy,
        )
    }

    /**
     * Скачать subscription URL + распарсить. Suspend — внутри переключается
     * на [Dispatchers.IO] потому что [HttpURLConnection] блокирующий.
     *
     * @param proxy если задан — fetch идёт через локальный HTTP-proxy
     *   (наш xray http-inbound при активном VPN). null = direct.
     *
     * Бросает [java.io.IOException] на сетевые проблемы (timeout, DNS,
     * 4xx/5xx). UI ловит и показывает.
     */
    suspend fun import(url: String, proxy: Proxy? = null): Result = withContext(Dispatchers.IO) {
        val rawConn = if (proxy != null) {
            URI(url).toURL().openConnection(proxy)
        } else {
            URI(url).toURL().openConnection()
        }
        val conn = (rawConn as HttpURLConnection).apply {
            connectTimeout = CONNECT_TIMEOUT_MS
            readTimeout = READ_TIMEOUT_MS
            requestMethod = "GET"
            instanceFollowRedirects = true
            setRequestProperty("User-Agent", USER_AGENT)
            // Provider'ы иногда возвращают gzip — JVM сама расжмёт если
            // accept-encoding не задан, но явно ставим identity для
            // предсказуемости (избегаем двойного decode).
            setRequestProperty("Accept-Encoding", "identity")
        }
        try {
            val code = conn.responseCode
            if (code !in 200..299) {
                throw java.io.IOException(
                    "HTTP $code от $url: ${conn.responseMessage}"
                )
            }
            val body = conn.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() }
            resultFromBody(body, viaProxy = proxy != null)
        } finally {
            conn.disconnect()
        }
    }

    /**
     * Импорт с автоматическим DPI-fallback. Pattern с десктопа
     * (`subscription.import_with_dpi_fallback`):
     *
     *   1. Пробуем direct fetch — fast path, не нагружаем туннель зря.
     *   2. Если упало с DPI-сигнатурой (TLS-handshake reset / EOF) И
     *      локальный xray-proxy слушает на [LOCAL_PROXY_HOST]:[LOCAL_PROXY_PORT]
     *      — retry через него. Подписка летит шифрованно через VPN-сервер,
     *      DPI не видит inner request.
     *   3. Если direct упал НЕ из-за DPI — пробрасываем оригинальное
     *      исключение (DNS, 4xx, timeout — не наша проблема).
     *   4. Если direct упал из-за DPI НО туннель не активен — тоже
     *      пробрасываем оригинал чтобы UI показал «подключись сначала».
     */
    suspend fun importWithDpiFallback(
        url: String,
        localProxyHost: String = LOCAL_PROXY_HOST,
        localProxyPort: Int = LOCAL_PROXY_PORT,
    ): Result {
        val directError: Throwable = try {
            return import(url)
        } catch (e: Throwable) { e }

        if (!looksLikeDpiBlock(directError)) {
            throw directError
        }
        if (!probeLocalProxy(localProxyHost, localProxyPort)) {
            // Туннель не активен — fallback некуда. Surface'им оригинальную
            // DPI-ошибку, UI говорит «подключись сначала».
            throw directError
        }
        val proxy = Proxy(Proxy.Type.HTTP, InetSocketAddress(localProxyHost, localProxyPort))
        return import(url, proxy = proxy)
    }

    /** Сигнатура российского DPI: TLS-handshake RST'ится middle of
     *  ClientHello. Наружу всплывает как [SSLException] / SocketException
     *  с фразами "reset" / "EOF" / etc. */
    internal fun looksLikeDpiBlock(err: Throwable): Boolean {
        // Type-based check — точнее чем строки. SSLException + любые wrapping.
        var e: Throwable? = err
        while (e != null) {
            if (e is SSLException) return true
            if (e is java.io.EOFException) return true
            e = e.cause
        }
        // Fallback на substring — wrapping в IOException делает type-check
        // недостаточным.
        val msg = err.message?.lowercase() ?: return false
        return msg.contains("connection reset") ||
            msg.contains("connection aborted") ||
            msg.contains("unexpected_eof") ||
            msg.contains("ssl handshake") ||
            msg.contains("remote host closed")
    }

    /** TCP-probe: что-то слушает [host]:[port]? Короткий timeout (500мс) —
     *  если xray не запущен, не хотим висеть. */
    internal fun probeLocalProxy(host: String, port: Int, timeoutMs: Int = 500): Boolean = try {
        Socket().use { sock ->
            sock.connect(InetSocketAddress(host, port), timeoutMs)
            true
        }
    } catch (_: Throwable) {
        false
    }

    /** Локальный HTTP-inbound xray. Совпадает с
     *  `XrayConfigBuilder.DEFAULT_LISTEN_HOST/PORT`. */
    private const val LOCAL_PROXY_HOST = "127.0.0.1"
    private const val LOCAL_PROXY_PORT = 2080
}

File: kapro_vpn/core/subscription.py
"""Download and parse VPN subscription URLs.

Most providers (dns.army, BMV, AmneziaFree, etc.) hand out a single URL
that returns a base64-encoded list of share-URLs — one per line. This
module fetches that, decodes it, and walks each line through the share-URL
parser to produce ProxyConfig objects ready for storage.

Format detection:
- Try base64-decode first; if the result starts with a known scheme
  (vless://, vmess://, trojan://, ss://, hysteria2://) treat as the
  intended payload.
- Otherwise assume the response is already plain text and parse directly.

DPI fallback:
- Many provider sites (gmailvpn.ru, getoutline.org mirrors, etc.) are
  blocked by Russian ISPs at the TLS layer — the TCP handshake completes
  but ClientHello gets RST'd before ServerHello. requests sees this as
  SSLEOFError / ConnectionResetError.
- When xray is already running locally, we can fetch the subscription
  *through* the active tunnel: route the HTTP request to xray's mixed
  inbound at 127.0.0.1:listen_port. Same trick the browser uses when
  the system proxy is set — DPI sees only the encrypted outbound xray
  stream and can't pattern-match the inner request.
"""
from __future__ import annotations

import base64
import socket
from dataclasses import dataclass, field
from typing import Optional

import requests

from .. import __version__
from .parser import ParseError, ProxyConfig, parse

# Self-identifying User-Agent so subscription providers have one obvious
# string to whitelist or block. Format follows the de-facto convention
# (`Name/Version (Platform; +URL)`) used by v2rayN, NekoBox, Streisand
# etc., so server-side allowlists that pattern-match "<Name>/" work out
# of the box.
USER_AGENT = f"KaproVPN/{__version__} (Windows; +https://github.com/fafnirov/KaproVPN)"

SUPPORTED_SCHEMES = ("vless://", "vmess://", "trojan://", "ss://",
                     "hysteria2://", "hy2://")


@dataclass
class SubscriptionResult:
    configs: list[ProxyConfig]
    errors: list[str]
    raw_lines: int  # how many candidate lines we tried to parse
    via_proxy: bool = False  # did we fall back to the local xray tunnel?
    # Names of configs that parsed fine but are provider "stubs" (e.g.
    # gmailvpn's `vless://…@0.0.0.0:1 #App not supported`) — filtered out
    # of `configs` so a dead placeholder is never imported as a server.
    placeholders: list[str] = field(default_factory=list)


@dataclass
class FetchError:
    """Structured classification of a subscription-fetch failure, so the
    UI can show an accurate cause instead of always blaming REALITY/DPI."""
    category: str        # not_found | auth | server | timeout | dpi | conn | unknown
    raw: str             # technical "TypeName: message" for the details line
    title: str           # short human-readable cause
    detail: str          # what the user should do
    suggest_manual: bool  # whether browser-copy-paste could plausibly help


# Hosts that mean "this isn't a real server" — providers hand these out as
# placeholders instead of a 404 when they won't serve a given client/plan.
PLACEHOLDER_HOSTS = {"", "0", "0.0.0.0", "127.0.0.1", "::", "::1", "localhost"}


def is_placeholder_config(cfg: ProxyConfig) -> bool:
    """True if a parsed config is a provider stub rather than a usable server.

    Catches the gmailvpn-style `vless://…@0.0.0.0:1 #App not supported`
    dummy: importing it would give the user a dead server with no hint why.
    """
    host = str(cfg.outbound.get("server", "")).strip().lower()
    if host in PLACEHOLDER_HOSTS:
        return True
    name = (cfg.name or "").lower()
    if "not supported" in name or "unsupported" in name:
        return True
    port = cfg.outbound.get("server_port", cfg.outbound.get("port"))
    try:
        if port is not None and int(port) <= 0:
            return True
    except (TypeError, ValueError):
        pass
    return False


def classify_fetch_error(exc: Exception) -> FetchError:
    """Map a fetch exception to an accurate, actionable message.

    The big win over the old "always REALITY/whitelist" hint: an HTTP
    status error means the server *answered* — it's not a TLS/DPI block,
    and manual browser-paste won't conjure a subscription that 404s.
    """
    raw = f"{type(exc).__name__}: {exc}"
    resp = getattr(exc, "response", None)
    status = getattr(resp, "status_code", None) if resp is not None else None
    if status is not None:
        if status in (404, 410):
            return FetchError(
                "not_found", raw,
                f"Подписка не найдена (HTTP {status}).",
                "Ссылка неверная или устарела — ручная вставка не поможет. "
                "Запроси у провайдера актуальную ссылку.",
                suggest_manual=False,
            )
        if status in (401, 403):
            return FetchError(
                "auth", raw,
                f"Доступ запрещён (HTTP {status}).",
                "Подписка не активна / требует авторизации, либо доступ "
                "ограничен по IP. Если URL открывается в браузере — "
                "скопируй ответ и вставь вручную.",
                suggest_manual=True,
            )
        return FetchError(
            "server", raw,
            f"Сервер провайдера ответил ошибкой (HTTP {status}).",
            "Это на стороне провайдера. Попробуй позже или запроси новую ссылку.",
            suggest_manual=False,
        )
    if _looks_like_dpi_block(exc):
        return FetchError(
            "dpi", raw,
            "Похоже на DPI-блокировку или REALITY / IP-белый список.",
            "Открой URL в браузере, скопируй ответ целиком и вставь вручную. "
            "Или подключись к любому серверу — KaproVPN попробует ещё раз "
            "через туннель.",
            suggest_manual=True,
        )
    if isinstance(exc, requests.exceptions.Timeout):
        return FetchError(
            "timeout", raw,
            "Таймаут — сервер не ответил.",
            "Проверь интернет и ссылку. Если сайт открывается в браузере — "
            "вставь ответ вручную.",
            suggest_manual=True,
        )
    if isinstance(exc, requests.exceptions.ConnectionError):
        return FetchError(
            "conn", raw,
            "Не удалось соединиться с сервером провайдера.",
            "Проверь ссылку и интернет. Если URL открывается в браузере — "
            "вставь ответ вручную.",
            suggest_manual=True,
        )
    return FetchError(
        "unknown", raw,
        "Не удалось загрузить подписку.",
        "Если URL открывается в браузере — скопируй ответ и вставь вручную.",
        suggest_manual=True,
    )


def parse_subscription_body(body: str) -> list[str]:
    """Pull share-URLs out of a subscription response body.

    Some providers ship plain text, some ship base64. We try base64 first
    if the body looks line-noise-y (no obvious share-URL anywhere), and
    fall back to the raw text otherwise.
    """
    body = body.strip()
    if not body:
        return []

    candidates = [body]
    # If the body doesn't have an obvious scheme already, try base64-decode
    if not any(sch in body for sch in SUPPORTED_SCHEMES):
        try:
            # base64 fix-padding: append '=' until length % 4 == 0
            padded = body + "=" * ((-len(body)) % 4)
            decoded = base64.b64decode(padded, validate=False).decode(
                "utf-8", errors="replace",
            )
            if any(sch in decoded for sch in SUPPORTED_SCHEMES):
                candidates.insert(0, decoded)
        except Exception:
            pass

    for text in candidates:
        urls = [
            line.strip() for line in text.splitlines()
            if line.strip()
            and not line.strip().startswith("#")
            and any(line.strip().startswith(sch) for sch in SUPPORTED_SCHEMES)
        ]
        if urls:
            return urls
    return []


def _fetch(url: str, timeout: tuple[float, float],
           proxy_url: Optional[str] = None) -> str:
    """One requests.get call; optionally routed through proxy_url."""
    proxies = None
    if proxy_url:
        # xray's mixed inbound speaks both HTTP and SOCKS on the same port,
        # so a single http://127.0.0.1:port URL handles http+https requests.
        proxies = {"http": proxy_url, "https": proxy_url}
    response = requests.get(url, timeout=timeout, proxies=proxies, headers={
        "User-Agent": USER_AGENT,
    })
    response.raise_for_status()
    return response.text


def _looks_like_dpi_block(err: Exception) -> bool:
    """Heuristic: did this failure look like a TLS-layer block?

    Russian DPI typically RSTs the connection mid-ClientHello, surfacing
    as SSLEOFError or a generic ConnectionResetError. ConnectionError with
    "EOF" / "reset" / "aborted" in the message is the same thing wrapped
    one layer up by urllib3.
    """
    msg = str(err).lower()
    needles = (
        "unexpected_eof_while_reading",
        "ssleoferror",
        "connection reset",
        "connectionreseterror",
        "connection aborted",
        "remoteend",
        "ssl: ",  # broad — covers misc handshake failures
    )
    return any(n in msg for n in needles)


def _probe_local_proxy(host: str, port: int, timeout: float = 0.5) -> bool:
    """True if something is listening at host:port (i.e. xray is up)."""
    try:
        with socket.create_connection((host, port), timeout=timeout):
            return True
    except OSError:
        return False


def result_from_body(body: str, via_proxy: bool = False) -> SubscriptionResult:
    """Take a raw subscription response and turn it into a SubscriptionResult.

    Shared between the URL-fetch flow and the manual-paste fallback in the
    UI, so a body the user copied from their browser parses identically
    to one we downloaded ourselves.
    """
    share_urls = parse_subscription_body(body)
    configs: list[ProxyConfig] = []
    errors: list[str] = []
    placeholders: list[str] = []
    for share_url in share_urls:
        try:
            cfg = parse(share_url)
        except ParseError as e:
            short = share_url[:60] + ("…" if len(share_url) > 60 else "")
            errors.append(f"{short} — {e}")
            continue
        if is_placeholder_config(cfg):
            placeholders.append(cfg.name or share_url[:40])
        else:
            configs.append(cfg)
    return SubscriptionResult(
        configs=configs, errors=errors, raw_lines=len(share_urls),
        via_proxy=via_proxy, placeholders=placeholders,
    )


def import_subscription(
    url: str,
    timeout: tuple[float, float] = (10, 20),
    proxy_url: Optional[str] = None,
) -> SubscriptionResult:
    """Download a subscription and parse every contained share-URL.

    `url` should be the provider-supplied subscription URL.
    `proxy_url`, if set, routes the fetch through it (e.g.
    "http://127.0.0.1:2080" to go via the active xray tunnel).
    Raises requests.RequestException on network failure.
    """
    body = _fetch(url, timeout, proxy_url=proxy_url)
    return result_from_body(body, via_proxy=bool(proxy_url))


def import_with_dpi_fallback(
    url: str,
    local_proxy_host: str = "127.0.0.1",
    local_proxy_port: int = 2080,
    timeout: tuple[float, float] = (10, 20),
) -> SubscriptionResult:
    """Fetch a subscription, automatically retrying via the local tunnel
    if the direct attempt looks DPI-blocked.

    Flow:
      1. Try a normal direct fetch first — fast and avoids loading the
         tunnel for no reason.
      2. On TLS-handshake-EOF / connection-reset (the Russian DPI
         signature), probe 127.0.0.1:listen_port. If xray is up, retry
         through it. The result's `via_proxy` flag tells the caller a
         fallback happened so they can surface it in the UI.
      3. Otherwise re-raise the original error untouched.
    """
    try:
        return import_subscription(url, timeout=timeout)
    except requests.RequestException as direct_err:
        if not _looks_like_dpi_block(direct_err):
            raise
        if not _probe_local_proxy(local_proxy_host, local_proxy_port):
            # No active tunnel to fall back through — surface the
            # original DPI error so the caller's "connect first" hint
            # makes sense to the user.
            raise
        proxy_url = f"http://{local_proxy_host}:{local_proxy_port}"
        # Let the proxied attempt's own exception escape unwrapped — if
        # it ALSO fails, that's more interesting than the DPI error we
        # already explained away.
        return import_subscription(url, timeout=timeout, proxy_url=proxy_url)
